### ISSUE TITLE
RPN output proposal's score

### ISSUE THREAD

**Issue by WEIJIE-XIE** [open-mmlab/mmdetection#5567]
# 1. Describe the issue
MMDetection set foreground (FG) label to 0 and  background (BG) label to 1 when training the RPN network.  

# 2. Results
As you will see from Section 3 (Detail), there is no ''Reverse transformation'' in the process of model training and testing. Theoretically, the smaller the output rpn_score is, the more likely it is to be FG (close to 0). However, the effect of show_box is the opposite. High-score boxes looks more like positive proposals, **because the green boxes are more centered around Ground Truth**.

**The bold blue box is ground truth. The high score box (theoretically BG, but looks more like FG) is green. The low score box (theoretically FG, but looks more like BG) is blue. **
![image1](https://i.loli.net/2021/07/08/pxQR4UiHDkKSZYX.png)
![image2](https://i.loli.net/2021/07/08/Ta1ZJkSKNzjm5sC.jpg)

# 3. Detail
more detail see in a blog (https://blog.csdn.net/beishang0031/article/details/111573020)

## label set

**mmdet/models/dense_heads/anchor_head.py, part of the function _get_targets_single()**
```
pos_inds = sampling_result.pos_inds
neg_inds = sampling_result.neg_inds
if len(pos_inds) > 0:
    if not self.reg_decoded_bbox:
        pos_bbox_targets = self.bbox_coder.encode(
            sampling_result.pos_bboxes, sampling_result.pos_gt_bboxes)
    else:
        pos_bbox_targets = sampling_result.pos_gt_bboxes
    bbox_targets[pos_inds, :] = pos_bbox_targets
    bbox_weights[pos_inds, :] = 1.0
    if gt_labels is None:
        # Only rpn gives gt_labels as None
        # Foreground is the first class since v2.5.0
        labels[pos_inds] = 0
    else:
        labels[pos_inds] = gt_labels[
            sampling_result.pos_assigned_gt_inds]
    if self.train_cfg.pos_weight <= 0:
        label_weights[pos_inds] = 1.0
    else:
        label_weights[pos_inds] = self.train_cfg.pos_weight
if len(neg_inds) > 0:
    label_weights[neg_inds] = 1.0

# map up to original set of anchors
if unmap_outputs:
    num_total_anchors = flat_anchors.size(0)
    labels = unmap(
        labels, num_total_anchors, inside_flags,
        fill=self.num_classes)  # fill bg label
    label_weights = unmap(label_weights, num_total_anchors,
                          inside_flags)
    bbox_targets = unmap(bbox_targets, num_total_anchors, inside_flags)
    bbox_weights = unmap(bbox_weights, num_total_anchors, inside_flags)
```

## train process
**model output, part of function AnchorHead.forword_single()**
```
self.rpn_conv = nn.Conv2d(
            self.in_channels, self.feat_channels, 3, padding=1)
self.rpn_cls = nn.Conv2d(self.feat_channels,
                                 self.num_anchors * self.cls_out_channels, 1)
self.rpn_reg = nn.Conv2d(self.feat_channels, self.num_anchors * 4, 1)

x = self.rpn_conv(x)
x = F.relu(x, inplace=True)
rpn_cls_score = self.rpn_cls(x)
rpn_bbox_pred = self.rpn_reg(x)
return rpn_cls_score, rpn_bbox_pred
```

**loss calculate，part of function AnchorHead.loss_single()**
```
# classification loss
labels = labels.reshape(-1)
label_weights = label_weights.reshape(-1)
cls_score = cls_score.permute(0, 2, 3,
                              1).reshape(-1, self.cls_out_channels)
loss_cls = self.loss_cls(
    cls_score, labels, label_weights, avg_factor=num_total_samples)
# regression loss
bbox_targets = bbox_targets.reshape(-1, 4)
bbox_weights = bbox_weights.reshape(-1, 4)
bbox_pred = bbox_pred.permute(0, 2, 3, 1).reshape(-1, 4)
if self.reg_decoded_bbox:
    # When the regression loss (e.g. `IouLoss`, `GIouLoss`)
    # is applied directly on the decoded bounding boxes, it
    # decodes the already encoded coordinates to absolute format.
    anchors = anchors.reshape(-1, 4)
    bbox_pred = self.bbox_coder.decode(anchors, bbox_pred)
loss_bbox = self.loss_bbox(
    bbox_pred,
    bbox_targets,
    bbox_weights,
    avg_factor=num_total_samples)
return loss_cls, loss_bbox
```

## test process
**predict proposal, part of function RPNHead._get_bboxes()**
```
rpn_cls_score = cls_scores[idx]
rpn_bbox_pred = bbox_preds[idx]
assert rpn_cls_score.size()[-2:] == rpn_bbox_pred.size()[-2:]
rpn_cls_score = rpn_cls_score.permute(0, 2, 3, 1)
if self.use_sigmoid_cls:
    rpn_cls_score = rpn_cls_score.reshape(batch_size, -1)
    scores = rpn_cls_score.sigmoid()
else:
    rpn_cls_score = rpn_cls_score.reshape(batch_size, -1, 2)
    # We set FG labels to [0, num_class-1] and BG label to
    # num_class in RPN head since mmdet v2.5, which is unified to
    # be consistent with other head since mmdet v2.0. In mmdet v2.0
    # to v2.4 we keep BG label as 0 and FG label as 1 in rpn head.
    scores = rpn_cls_score.softmax(-1)[..., 0]
rpn_bbox_pred = rpn_bbox_pred.permute(0, 2, 3, 1).reshape(
    batch_size, -1, 4)
anchors = mlvl_anchors[idx]
anchors = anchors.expand_as(rpn_bbox_pred)
nms_pre = cfg.get('nms_pre', -1)
if nms_pre > 0 and rpn_bbox_pred.size(1) > nms_pre:
    # sort is faster than topk
    ranked_scores, rank_inds = scores.sort(descending=True)
    topk_inds = rank_inds[:, :cfg.nms_pre]
    scores = ranked_scores[:, :cfg.nms_pre]
    batch_inds = torch.arange(batch_size).view(
        -1, 1).expand_as(topk_inds)
    rpn_bbox_pred = rpn_bbox_pred[batch_inds, topk_inds, :]
    anchors = anchors[batch_inds, topk_inds, :]

mlvl_scores.append(scores)
mlvl_bbox_preds.append(rpn_bbox_pred)
mlvl_valid_anchors.append(anchors)
level_ids.append(
    scores.new_full((
        batch_size,
        scores.size(1),
    ),
                    idx,
                    dtype=torch.long))
```

# 4. Issue fix

If you have already identified the reason, you can provide the information here. If you are willing to create a PR to fix it, please also leave a comment here and that would be much appreciated!

**Comment by jshilong** [comment]
We use 0 to represent foreground the `num_classes` to represent the background from mmdet 2.0
You can refer to  https://github.com/open-mmlab/mmdetection/blob/master/docs/compatibility.md

**Comment by jshilong** [comment]
Feel free to reopen the issue if there is any question
